util/progress/progressui/display.go

```diff
@@ -71,53 +83,152 @@ func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch
+	return &consoleDisplay{
+		t:    newTrace(c, true),
+		disp: &display{c: c},
+	}
+}
+
+func (d *consoleDisplay) init(dsso *displaySolveStatusOpts) {
+	d.disp.phase = dsso.phase
+	d.disp.desc = dsso.consoleDesc
+	if d.disp.phase == "" {
+		d.disp.phase = "Building"
+	}
+	d.width, d.height = d.disp.getSize()
+}
+
+func (d *consoleDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, d.width)
+}
+
+func (d *consoleDisplay) display(done bool) {
+	d.width, d.height = d.disp.getSize()
+	d.disp.print(d.t.displayInfo(), d.width, d.height, done)
+	if done {
+		d.t.printErrorLogs(d.t.w)
+	}
+}
+
+func (d *consoleDisplay) warnings() []client.VertexWarning {
+	return d.t.warnings()
+}
+
+type textDisplay struct {
+	t       *trace
+	printer *textMux
+}
+
+func newTextDisplay(w io.Writer, printer vertexPrinter) SolveStatusDisplay {
+	return &textDisplay{
+		t:       newTrace(w, false),
+		printer: newTextMux(w, printer),
+	}
+}
+
+func newJSONDisplay(w io.Writer) SolveStatusDisplay {
+	return newTextDisplay(w, &jsonPrinter{})
+}
+
+func (d *textDisplay) init(dsso *displaySolveStatusOpts) {
+	d.printer.desc = dsso.textDesc
+}
+
+func (d *textDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, 80)
+}
+
+func (d *textDisplay) display(done bool) {
+	d.printer.print(d.t)
+	if done {
+		d.t.printErrorLogs(d.t.w)
+	}
+}
+
+func (d *textDisplay) warnings() []client.VertexWarning {
+	return d.t.warnings()
+}
+
+type DisplayMode string
+
+const (
+	AutoMode  DisplayMode = "auto"
+	TtyMode   DisplayMode = "tty"
+	PlainMode DisplayMode = "plain"
+	JSONMode  DisplayMode = "json"
+)
+
+// NewDisplay constructs a SolveStatusDisplay that outputs to the given console.File with the given DisplayMode.
+//
+// This method will return an error when the DisplayMode is invalid or if TtyMode is used but the console.File
+// does not refer to a tty. AutoMode will choose TtyMode or PlainMode depending on if the output is a tty or not.
+func NewDisplay(out console.File, mode DisplayMode) (SolveStatusDisplay, error) {
+	switch mode {
+	case AutoMode, TtyMode, "":
+		if c, err := console.ConsoleFromFile(out); err == nil {
+			return NewConsoleDisplay(c), nil
+		} else if mode == "tty" {
+			return nil, errors.Wrap(err, "failed to get console")
+		}
+		fallthrough
+	case PlainMode:
+		return newTextDisplay(out, nil), nil
+	case JSONMode:
+		return newJSONDisplay(out), nil
+	default:
+		return nil, errors.Errorf("invalid progress mode %s", mode)
+	}
+}
+
+// DisplaySolveStatus will display solve status updates to the given console or writer.
+// If console.Console is nil, this outputs status updates in the plain text format to the io.Writer.
+//
+// For more control over the display method, it is suggested to use NewDisplay and UpdateDisplay instead
+// of this function.
+func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch chan *client.SolveStatus, opts ...DisplaySolveStatusOpt) ([]client.VertexWarning, error) {
```

Comment:
I think this is unused now. If correct then we can remove it and use `d = NewDisplay(...); d.Show(ch)`. (I think it might make sense to make `d` and the backend interface for display different objects then).

Reply:
I think we can mark this function as deprecated. I don't think we should remove it in the same commit as adding a new method. It would still be used by other repositories that aren't connected to this one. Those should be given an opportunity to be able to migrate to the new API.

Reply:
We don't have backward compatibility guarantees for Go vendors. This will just add maintenance cost(and stops us from making changes if we need to). Importers will still need to update the code anyway. Marking function deprecated usually means nobody will remember to remove it.